server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/ComponentTreeBuilder.java

```diff
@@ -0,0 +1,209 @@
+    this.scannerComponentSupplier = scannerComponentSupplier;
+    this.project = project;
+    this.baseAnalysis = baseAnalysis;
+    this.skipUnchangedFiles = skipUnchangedFiles;
+  }
+
+  public Component buildProject(ScannerReport.Component project) {
+    return buildComponent(project, project);
+  }
+
+  private Component[] buildChildren(ScannerReport.Component component, ScannerReport.Component parentModule) {
+    return component.getChildRefList()
+      .stream()
+      .map(scannerComponentSupplier::apply)
+      .filter(c -> !skipUnchangedFiles || c.getStatus() != FileStatus.SAME)
+      .map(c -> buildComponent(c, parentModule))
+      .toArray(Component[]::new);
+  }
+
+  private ComponentImpl buildComponent(ScannerReport.Component component, ScannerReport.Component closestModule) {
+    switch (component.getType()) {
+      case PROJECT:
+        String projectKey = keyGenerator.generateKey(component, null);
+        String uuid = uuidSupplier.apply(projectKey);
+        return ComponentImpl.builder(Component.Type.PROJECT)
+          .setUuid(uuid)
+          .setKey(projectKey) 
+          .setName(nameOfProject(component))
+          .setStatus(convertStatus(component.getStatus()))
+          .setDescription(trimToNull(component.getDescription()))
+          .setReportAttributes(createAttributesBuilder(component)
+            .setVersion(createProjectVersion(component))
+            .build())
+          .addChildren(buildChildren(component, component))
+          .build();
+
+      case MODULE:
+        String moduleKey = keyGenerator.generateKey(component, null);
+        return ComponentImpl.builder(Component.Type.MODULE)
+          .setUuid(uuidSupplier.apply(moduleKey))
+          .setKey(moduleKey)
+          .setName(nameOfOthers(component, moduleKey))
+          .setStatus(convertStatus(component.getStatus()))
+          .setDescription(trimToNull(component.getDescription()))
+          .setReportAttributes(createAttributesBuilder(component).build())
+          .addChildren(buildChildren(component, component))
+          .build();
+
+      case DIRECTORY:
+      case FILE:
+        String key = keyGenerator.generateKey(closestModule, component);
+        return ComponentImpl.builder(convertDirOrFileType(component.getType()))
+          .setUuid(uuidSupplier.apply(key))
+          .setKey(key)
+          .setName(nameOfOthers(component, key))
+          .setStatus(convertStatus(component.getStatus()))
+          .setDescription(trimToNull(component.getDescription()))
+          .setReportAttributes(createAttributesBuilder(component).build())
+          .setFileAttributes(createFileAttributes(component))
+          .addChildren(buildChildren(component, closestModule))
+          .build();
+
+      default:
+        throw new IllegalArgumentException(format("Unsupported component type '%s'", component.getType()));
+    }
+  }
+  
+  private static Component.Status convertStatus(ScannerReport.Component.FileStatus status) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace this fully qualified name with "FileStatus" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1942)